truffle/com.oracle.truffle.api.instrumentation/src/com/oracle/truffle/api/instrumentation/SourceSectionFilter.java

```diff
@@ -0,0 +1,540 @@
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package com.oracle.truffle.api.instrumentation;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+
+import com.oracle.truffle.api.source.Source;
+import com.oracle.truffle.api.source.SourceSection;
+
+/**
+ * A source section filter represents an expression for a subset of tagged source sections that are
+ * used in an Truffle interpreter.
+ * <p>
+ * Start building event filters by calling {@link SourceSectionFilter#newBuilder()} and complete
+ * them by calling {@link Builder#build()}.
+ *
+ * @see SourceSectionFilter#newBuilder()
+ * @see Instrumenter#attachFactory(SourceSectionFilter, EventNodeFactory)
+ * @see Instrumenter#attachListener(SourceSectionFilter, EventListener)
+ */
+public final class SourceSectionFilter {
+
+    private final EventFilterExpression[] nodeExpressions;
+    private final EventFilterExpression[] rootNodeExpressions;
+
+    private SourceSectionFilter(EventFilterExpression[] rootNodeExpressions, EventFilterExpression[] nodeExpressions) {
+        this.rootNodeExpressions = rootNodeExpressions;
+        this.nodeExpressions = nodeExpressions;
+    }
+
+    /**
+     * Creates a new {@link SourceSectionFilter} expression using a {@link Builder builder} pattern.
+     * Individual builder statements are interpreted as conjunctions (AND) while multiple parameters
+     * for individual filter expressions are treated as disjunctions (OR). To create the final
+     * filter finalize the expression using {@link Builder#build()}.
+     *
+     * @see Builder#sourceIs(Source...)
+     * @see Builder#mimeTypeIs(String...)
+     * @see Builder#tagIs(String...)
+     * @see Builder#tagIsNot(String...)
+     * @see Builder#sourceSectionEquals(SourceSection...)
+     * @see Builder#indexIn(int, int)
+     * @see Builder#lineIn(int, int)
+     * @see Builder#lineIs(int)
+     * @see Builder#build()
+     *
+     * @return a new builder to create
+     */
+    public static Builder newBuilder() {
+        return new Builder();
+    }
+
+    public static final class Builder {
+
+        private List<EventFilterExpression> nodeExpressions = new ArrayList<>();
+        private List<EventFilterExpression> rootNodeExpressions = new ArrayList<>();
+
+        /**
+         * Add a filter for all source sections that reference one of the given sources.
+         */
+        public Builder sourceIs(Source... source) {
+            verifyNotNull(source);
+            rootNodeExpressions.add(new EventFilterExpression.SourceIs(source));
+            return this;
+        }
+
+        /**
+         * Add a filter for all source sections that declare one of the given mime-types.
```

Comment:
How are MIME types matched?  Jarda has written that all MIME types supported by a `Language` are synonyms.  Will this filter match synonymous MIME types?  If not, how might a client discover which synonym to use?